tests/queries/0_stateless/02354_annoy_index.sql

```diff
@@ -1,150 +1,251 @@
 
--- first argument (distance_function) must be String
-CREATE TABLE tab(id Int32, embedding Array(Float32), INDEX annoy_index embedding TYPE annoy(3)) ENGINE = MergeTree ORDER BY id; -- { serverError INCORRECT_QUERY }
+-- rows = 16, index_granularity = 4, GRANULARITY = 4 gives a single annoy-indexed block (comprising all granules)
+-- no two matches happen to be located in the same granule, so with LIMIT = 3, we'll get "Granules: 2/4"
 
--- 2nd argument (number of trees) must be UInt64
-CREATE TABLE tab(id Int32, embedding Array(Float32), INDEX annoy_index embedding TYPE annoy('L2Distance', 'not an UInt64')) ENGINE = MergeTree ORDER BY id; -- { serverError INCORRECT_QUERY }
+-- See (*) why commented out
+-- SELECT 'WHERE type, L2Distance';
+-- SELECT *
+-- FROM tab
+-- WHERE L2Distance(vector, [10.0, 0.0, 10.0, 0.0]) < 5.0
+-- LIMIT 3;
 
--- reject unsupported distance functions
-CREATE TABLE tab(id Int32, embedding Array(Float32), INDEX annoy_index embedding TYPE annoy('wormholeDistance')) ENGINE = MergeTree ORDER BY id; -- { serverError INCORRECT_DATA }
+SELECT 'WHERE type, L2Distance, check that index is used';
+EXPLAIN indexes=1
+SELECT *
+FROM tab
+WHERE L2Distance(vector, [10.0, 0.0, 10.0, 0.0]) < 5.0
+LIMIT 3;
 
--- must be created on single column
-CREATE TABLE tab(id Int32, embedding Array(Float32), INDEX annoy_index (embedding, id) TYPE annoy()) ENGINE = MergeTree ORDER BY id; -- { serverError INCORRECT_NUMBER_OF_COLUMNS }
+-- See (*) why commented out
+-- SELECT 'ORDER BY type, L2Distance';
+-- SELECT *
+-- FROM tab
+-- ORDER BY L2Distance(vector, [10.0, 0.0, 10.0, 0.0])
+-- LIMIT 3;
 
--- must be created on Array/Tuple(Float32) columns
-SET allow_suspicious_low_cardinality_types = 1;
-CREATE TABLE tab(id Int32, embedding Float32, INDEX annoy_index embedding TYPE annoy()) ENGINE = MergeTree ORDER BY id; -- { serverError ILLEGAL_COLUMN }
-CREATE TABLE tab(id Int32, embedding Array(Float64), INDEX annoy_index embedding TYPE annoy()) ENGINE = MergeTree ORDER BY id; -- { serverError ILLEGAL_COLUMN }
-CREATE TABLE tab(id Int32, embedding LowCardinality(Float32), INDEX annoy_index embedding TYPE annoy()) ENGINE = MergeTree ORDER BY id; -- { serverError ILLEGAL_COLUMN }
-CREATE TABLE tab(id Int32, embedding Nullable(Float32), INDEX annoy_index embedding TYPE annoy()) ENGINE = MergeTree ORDER BY id; -- { serverError ILLEGAL_COLUMN }
+SELECT 'ORDER BY type, L2Distance, check that index is used';
+EXPLAIN indexes=1
+SELECT *
+FROM tab
+ORDER BY L2Distance(vector, [10.0, 0.0, 10.0, 0.0])
+LIMIT 3;
+
+DROP TABLE tab;
+
+-- (*) Storage and search in Annoy indexes is inherently random. Tests which check for exact row matches would be unstable. Therefore,
```

Comment:
AFAIS, Annoy implementation supports `void set_seed(R q)` method. Probably, we could use to generate repeatable index. So, logically, using it and disabling merges can allow getting stable results for the queries